test/functional/test-xhr.js

```diff
@@ -105,6 +117,27 @@ describe('XHR', function() {
         expect(requests[0].method).to.equal('GET');
         expect(requests[1].method).to.equal('POST');
       });
+
+      it('should inject source origin query parameter', () => {
+        setupMockXhr();
+        xhr.fetchJson('/get?k=v1#h1');
+        expect(noOrigin(requests[0].url)).to.equal(
+            '/get?k=v1&__amp_source_origin=https%3A%2F%2Facme.com#h1');
+      });
+
+      it('should inject source origin query parameter w/o query', () => {
+        setupMockXhr();
+        xhr.fetchJson('/get');
+        expect(noOrigin(requests[0].url)).to.equal(
+            '/get?__amp_source_origin=https%3A%2F%2Facme.com');
+      });
+
+      it('should defend against invalid source origin query parameter', () => {
+        setupMockXhr();
+        expect(() => {
+          xhr.fetchJson('/get?k=v1&__amp_source_origin=invalid#h1');
```

Comment:
Also add test case where __amp_source_origin appears not in query string and for both where part of it is percent encoded.